docs/features/scripting.rst

```diff
@@ -96,29 +118,405 @@ where:
+
+Even though this is a fully qualified CMD2 error the py_script must look for this error and perform error checking.::
+
+  app('speak')
+  print("Working")
+
+::
+
+  (Cmd) run_pyscript script.py 
+  Working
+  (Cmd) 
+
+You should notice that no error message is printed. Let's utilize the ``CommandResult`` 
+object to inspect the actual returned data.::
+
+  result = app('speak')
+  print(result)
+::
+
+  (Cmd) run_pyscript script.py 
+  CommandResult(stdout='', stderr='Usage: speak [-h] [-p] [-s] [-r REPEAT] words [...]\nError: the following arguments are required: words\n\n', stop=False, data=None)
+
+Now we can see that there has been an error. Let's re write the script to perform error checking.::
+
+  result = app('speak')
+
+  if result.stderr:
+      print("Something went wrong")
+
+::
+
+  (Cmd) run_pyscript script.py 
+  Something went wrong
+
+In python development is good practice to fail and exit quickly after user input.::
+
+  import sys
+
+  result = app('speak TRUTH!!')
+
+  if result.stderr:
+      print("Something went wrong")
+      sys.exit()
+
+  print("Continuing along..")
+
+::
+
+  (Cmd) run_pyscript script.py 
+  Continuing along..
+
+We changed the input to be a valid ``speak`` command but no output. Again we must inspect the 
+``CommandResult``::
+
+  import sys
+
+  #Syntax error
+  result = app('speak TRUTH!!!')
+  if result.stderr:
+      print("Something went wrong")
+      sys.exit()
+
+  print(result.stdout)
+
+::
+
+  (Cmd) run_pyscript script.py 
+  TRUTH!!!
+
+By just using ``stdout`` and ``stderr`` it is possible to string together commands
+with rudimentary control flow. In the next section we will show how to take advantage of
+cmd_result data.
+
+Developing an Advanced API
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Until now the application designer has paid little attention to scripters and their needs.
+Wouldn't it be nice if while creating py_scripts one did not have to parse data from ``stdout``? We can 
+accomodate the weary scripter by adding one small line at the end of our ``do_*`` commands.
+
+``self.last_result = <value>``
+
+Adding the above line supercharges a cmd2 application and opens a new world of possibilities.
+
+.. note::
+
+  When setting results for a command function inside of a CommandSet use the private cmd instance::
+
+    self._cmd.last_result = <value>
+
+
+In the following command example we return an array containing directory elements.::
+
+    dir_parser = cmd2.Cmd2ArgumentParser()
+    dir_parser.add_argument('-l', '--long', action='store_true', help="display in long format with one item per line")
+
+    @cmd2.with_argparser(dir_parser, with_unknown_args=True)
+    def do_dir(self, args, unknown):
+        """List contents of current directory."""
+        # No arguments for this command
+        if unknown:
+            self.perror("dir does not take any positional arguments:")
+            self.do_help('dir')
+            self.last_result = 'Bad arguments'
```

Comment:
I'd probably set `last_result` to `None` since the presence of the error in `CommandResult.stderror` already supports using the `CommandResult` object as a bool to check for errors.

Since `CommandResult.__bool__()` gives preference to a non-None `data` member, it will treat "Bad arguments" as a boolean True value.

Reply:
Is last_result set to None automatically?  The way I've been  writing commands is to set last_result iff there is something positive to return.

It makes it easier for the designer.